### PR TITLE
Extract slots with mapping conditions during form activation

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- Fixes the extraction of a slot with mapping conditions that should be filled with information from the user message that triggers the form.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/actions/forms.py
Comment (m-vdb) [inline rasa/core/actions/forms.py:632]: would be rad to be able to do compare events with `!=` or `==` 🤔 I thought this was possible?

Comment (m-vdb) [reply rasa/core/actions/forms.py:632]: 🤔 I see that `__eq__` raises `NotImplemented`, which is really not helpful ahah

File: rasa/core/actions/forms.py
Comment (m-vdb) [inline rasa/core/actions/forms.py:589]: should we update the docstring of the function?

File: tests/core/actions/test_forms.py
Comment (m-vdb) [inline tests/core/actions/test_forms.py:102]: should the test be duplicated instead?

File: rasa/core/actions/forms.py
Comment (m-vdb) [inline rasa/core/actions/forms.py:711]: I am not sure what this function does, especially the operation on the temp tracker. Can you explain?

**Review by m-vdb — COMMENTED** [review]
Nice turnaround @ancalita 🚀 I've left a few comments

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py:606]: Since the tracker that's used effectively is a temporary tracker during `action.run()`, it doesn't matter much that the log gets issued after this update. I will explain in your other comment the decision behind this temp tracker.

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py:608]: I have to update the tracker immediately in order for the tracker to update its `tracker.slots` attribute, which gets used in the for loop below (L610-612) to extract the prefilled slots. Because the tracker is temporary, it won't duplicate the events in the **actual** tracker that the processor updates.

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py:589]: Yup, I'll see how I can add more context to the docstring about this new code.

File: tests/core/actions/test_forms.py
Comment (ancalita) [reply tests/core/actions/test_forms.py:102]: Yep, good point 👍🏼

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py:711]: If you look into the `LoopAction.run()` which gets inherited by `FormAction`, this method does three things:
1. activates the form
2. allows the form to collect info (or `do`)
3. deactivates the form once conditions for deactivation are met

It is only after all these steps are done, does the MessageProcessor update the tracker with events (which includes the vital `ActiveLoop` event which sets the `tracker.active_loop_name` to the form name). This is what gets checked by the default action `action_extract_slots` [here](https://github.com/RasaHQ/rasa/blob/e8c0f8495a3960a0feac64fbf3f2fb1c75cc49ce/rasa/core/actions/action.py#L1028) when coming across slots with mapping conditions.

Because this update of the tracker by the MessageProcessor happens **too late**, I had to override a part of the superclass `run` method, namely the `activation` part which I extracted and used a temp tracker instead that gets updated with `ActiveLoop` event in L707 - this is a default activation event. This way ACTION_EXTRACT_SLOTS can fill a slot with a mapping condition during activation, because the `temp_tracker.active_loop_name` will be set to the current form name and will pass the check in the default action `_verify_mapping_conditions()`.

Let me know if any of this is not clear enough 🙏🏼

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py:632]: I think implementing `__eq__` in events is out of scope here 😬 Happy to submit a ticket though for a clean-up?

File: tests/core/actions/test_forms.py
Comment (ancalita) [reply tests/core/actions/test_forms.py:175]: done: https://github.com/RasaHQ/rasa/pull/11149/commits/8c87ff1eac90aca80a73d27c3d56afbb8e9bae74

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py]: Done: https://github.com/RasaHQ/rasa/pull/11149/commits/8c87ff1eac90aca80a73d27c3d56afbb8e9bae74

File: rasa/core/actions/forms.py
Comment (m-vdb) [reply rasa/core/actions/forms.py:631]: yeah of course. I am not sure how much it's worth it and how much it would simplify the code, maybe that's worth having a technical spike for that 🤔

Comment (ancalita) [reply rasa/core/actions/forms.py:631]: Opened a [spike ticket](https://rasahq.atlassian.net/browse/ATO-128)

File: rasa/core/actions/forms.py
Comment (m-vdb) [reply rasa/core/actions/forms.py:710]: OK I think that I understand, thanks a lot for detailing this. Something that was confusing to was the use of a temp tracker. Basically it is mainly used here as a way to compute events that need to actually be set in the real tracker.

File: rasa/core/actions/loops.py
Comment (m-vdb) [inline rasa/core/actions/loops.py:102]: can this line actually stay in `def run()`?

Comment (ancalita) [reply rasa/core/actions/loops.py:102]: Yup, of course, less duplication 😄

**Comment by m-vdb** [comment]
@ancalita given Arjaan response on the ticket, I think this should wait a review from Lukas and/or Melinda. I initially started giving a review in order to go faster but now we're waiting on customer feedback.

**Comment by joeri-jongen-kbc-be** [comment]
> @ancalita given Arjaan response on the ticket, I think this should wait a review from Lukas and/or Melinda. I initially started giving a review in order to go faster but now we're waiting on customer feedback.

@m-vdb @ancalita 
I'm working for the customer, we validated the fix and it seems to work 🥳

**Review by losterloh — APPROVED** [review]
Very nice 🚀  Just left some small comments and will leave it up to you whether to tackle them or not.

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py]: Done: https://github.com/RasaHQ/rasa/pull/11149/commits/fbe279cc616fed9887016b21a33f374da7f1519b

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py]: Done: https://github.com/RasaHQ/rasa/pull/11149/commits/fbe279cc616fed9887016b21a33f374da7f1519b

File: rasa/core/actions/forms.py
Comment (ancalita) [reply rasa/core/actions/forms.py]: In the end I opted for a slightly different solution by removing the events as a parameter but I think it addresses your concerns, let me know wdyt: https://github.com/RasaHQ/rasa/pull/11149/commits/fbe279cc616fed9887016b21a33f374da7f1519b

**Review by losterloh — APPROVED** [review]

**Pull request merged**
